server/sonar-server/src/main/java/org/sonar/server/batch/IssuesAction.java

```diff
@@ -77,38 +78,53 @@ public void define(WebService.NewController controller) {
 
   @Override
   public void handle(Request request, Response response) throws Exception {
-    response.stream().setMediaType(MediaTypes.PROTOBUF);
-
-    try (DbSession session = dbClient.openSession(false)) {
+    try (DbSession dbSession = dbClient.openSession(false)) {
       String componentKey = request.mandatoryParam(PARAM_KEY);
-      ComponentDto component = componentFinder.getByKey(session, componentKey);
+      ComponentDto component = componentFinder.getByKey(dbSession, componentKey);
       userSession.checkComponentPermission(USER, component);
 
-      Map<String, String> keysByUUid = keysByUUid(session, component);
+      Map<String, String> keysByUUid = keysByUUid(dbSession, component);
+
+      ScannerInput.ServerIssue.Builder responseBuilder = ScannerInput.ServerIssue.newBuilder();
+      response.stream().setMediaType(MediaTypes.PROTOBUF);
+      OutputStream output = response.stream().output();
 
-      ScannerInput.ServerIssue.Builder issueBuilder = ScannerInput.ServerIssue.newBuilder();
-      for (Iterator<IssueDoc> issueDocIterator = issueIndex.selectIssuesForBatch(component); issueDocIterator.hasNext();) {
-        handleIssue(issueDocIterator.next(), issueBuilder, keysByUUid, response.stream().output());
+      ResultHandler<IssueDto> handler = resultContext -> {
+        IssueDto issue = resultContext.getResultObject();
+        handleIssue(issue, responseBuilder, keysByUUid, output);
+      };
+      switch (component.scope()) {
+        case Scopes.PROJECT:
+          dbClient.issueDao().scrollNonClosedByModuleOrProject(dbSession, component, handler);
+          break;
+        case Scopes.FILE:
+          dbClient.issueDao().scrollNonClosedByComponentUuid(dbSession, component.uuid(), handler);
+          break;
+        default:
+          // only projects, modules and files are supported. Other types of components are not allowed.
+          throw new IllegalStateException(format("Component of scope '%s' is not allowed", component.scope()));
```

Comment:
Correct. I will change (even if it's out-of-scope of this PR).
@henryju @dbmeneses @janos-ss is it a problem for SonarLint if WS batch/issues returns 400 instead of 500 when requested component is not a project, module or file ?